test/functional/sessions_controller_test.rb

```diff
@@ -78,6 +82,35 @@ class SessionsControllerTest < ActionController::TestCase
         end
       end
     end
+
+    context "when OTP is correct but session expired" do
+      setup do
+        @controller.session[:mfa_user] = @user.id
+        @controller.session[:mfa_expires_at] = 15.minutes.from_now
+        travel 30.minutes
```

Comment:
You need to "travel back" as well or use the block form (`travel` method).

My intention was to stop changing session manually (since it should be done in controller). So instead of changing `session[:mfa_expires_at]` directly, you can:

1. login and get redirected to OTP page
2. visit the otp page -> that should trigger the controller part setting `session[:mfa_expires_at]`
3. travel to the future and submit the login form (doing the post request)
4. ensure it failed to login

That way this test would reflect the "real-word" flow.